File: Java/01.ActiveMq.Verify/repository-implementations/src/main/java/org/cqrs101/HsqlDbRepositoryHelper.java
package org.cqrs101;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.UUID;
import java.util.function.BiConsumer;
import java.util.function.Function;

public class HsqlDbRepositoryHelper implements RepositoryHelper {

    private static final ObjectMapper mapper = new ObjectMapper();
    private String name;
    private final String driverName = "org.hsqldb.jdbcDriver";
    private Class clazz;

    private Connection createConnection() throws SQLException {
        return DriverManager.getConnection("jdbc:hsqldb:hsql://localhost/testdb", "SA", "");
    }

    @Override
    public RepositoryHelper create(Class clazz) {
        Connection conn = null;
        this.clazz = clazz;
        try {
            Class.forName(driverName);
            HsqlDbRepositoryHelper helper = new HsqlDbRepositoryHelper();
            helper.name = clazz.getSimpleName().toUpperCase(Locale.ROOT);

            conn = createConnection();
            Statement stmt = conn.createStatement();
            stmt.executeUpdate(
                    "CREATE TABLE " + helper.name + " ("
                    + "id UUID  NOT NULL, data VARCHAR(MAX) NOT NULL,,"
                    + "PRIMARY KEY (id));");
            return helper;
        } catch (Exception ex) {
            throw new RuntimeException("Missing hslqDb Driver", ex);
        } finally {
            try {
                if (conn != null) {
                    conn.close();
                }
            } catch (SQLException ex) {
                System.out.println(ex);
            }
        }
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public Object getById(UUID id) {
        Connection conn = null;
        try {
            conn = createConnection();
            Statement stmt = conn.createStatement();
            ResultSet resultSet = stmt.executeQuery(
                    "SELECT data FROM " + this.name + " WHERE "
                    + "id ='" + id + "';");
            for (; resultSet.next();) {
                String data = resultSet.getString("data");
                return mapper.readValue(data, clazz);
            }
            return null;
        } catch (Exception ex) {
            throw new RuntimeException("Missing hslqDb Driver", ex);
        } finally {
            try {
                if (conn != null) {
                    conn.close();
                }
            } catch (SQLException ex) {
                System.out.println(ex);
            }
        }
    }

    @Override
    public List<Object> getAll() {
        Connection conn = null;
        try {
            List<Object> result = new ArrayList<Object>();
            conn = createConnection();
            Statement stmt = conn.createStatement();
            ResultSet resultSet = stmt.executeQuery(
                    "SELECT * FROM " + this.name + ";");
            for (; resultSet.next();) {
                String data = resultSet.getString("data");
                Object target = mapper.readValue(data, clazz);
                result.add(target);
            }
            return result;
        } catch (Exception ex) {
            throw new RuntimeException("Missing hslqDb Driver", ex);
        } finally {
            try {
                if (conn != null) {
                    conn.close();
                }
            } catch (SQLException ex) {
                System.out.println(ex);
            }
        }
    }

    @Override
    public Object save(Object item, BiConsumer<Object, UUID> idSetter, Function<Object, UUID> idGetter) {
        Connection conn = null;
        try {
            boolean exists = false;
            conn = createConnection();
            UUID id = idGetter.apply(item);
            if (id != null) {
                Statement stmt = conn.createStatement();
                ResultSet resultSet = stmt.executeQuery(
                        "SELECT data FROM " + this.name + " WHERE "
                        + "id ='" + id + "';");
                for (; resultSet.next();) {
                    exists = true;
                    break;
                }
            } else {
                id = UUID.randomUUID();
                idSetter.accept(item, id);
            }
            String itemString = mapper.writeValueAsString(item);
            Statement stmt = conn.createStatement();
            if (!exists) {
                stmt.execute(
                        "INSERT INTO " + this.name + "(id,data) VALUES ("
                        + "'" + id + "','" + itemString + "';");
            } else {
                stmt.execute(
                        "UPDATE " + this.name + " set data = '" + itemString + "'"
                        + "WHERE id ='" + id + "';");
            }
            return item;
        } catch (Exception ex) {
            throw new RuntimeException("Missing hslqDb Driver", ex);
        } finally {
            try {
                if (conn != null) {
                    conn.close();
                }
            } catch (SQLException ex) {
                System.out.println(ex);
            }
        }
    }

    @Override
    public void delete(UUID id) {
        Connection conn = null;
        try {
            conn = createConnection();
            Statement stmt = conn.createStatement();
            stmt.execute(
                    "DELETE FROM " + this.name + " WHERE "
                    + "id ='" + id + "';");
        } catch (Exception ex) {
            throw new RuntimeException("Missing hslqDb Driver", ex);
        } finally {
            try {
                if (conn != null) {
                    conn.close();
                }
            } catch (SQLException ex) {
                System.out.println(ex);
            }
        }
    }
}

File: Net/Commons/Repository.Implementations/HsqlDbRepositoryHelper.cs
using Repository.Interfaces;
using System;
using System.Linq;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Text;

namespace Repository.Implementations
{
    public class HsqlDbRepositoryHelper : IRepositoryHelper
    {
        private static ConcurrentDictionary<Type, ConcurrentDictionary<Guid, Object>> storage = new ConcurrentDictionary<Type, ConcurrentDictionary<Guid, object>>();
        private Type clazz;
        
        public HsqlDbRepositoryHelper()
        {
            throw new NotImplementedException();
        }

        public IRepositoryHelper Create(Type clazz)
        {
            storage[clazz] = new ConcurrentDictionary<Guid, Object>();
            HsqlDbRepositoryHelper helper = new HsqlDbRepositoryHelper();
            helper.clazz = clazz;
            helper.Name = clazz.Name;
            return helper;
        }


        public String Name { get; private set; }


        public Object GetById(Guid id)
        {
            ConcurrentDictionary<Guid, Object> data = storage[clazz];
            if (!data.ContainsKey(id)) return null;
            return data[id];
        }


        public List<Object> GetAll()
        {
            ConcurrentDictionary<Guid, Object> data = storage[clazz];
            return data.Values.ToList();
        }


        public Object Save(Object item, Action<Object, Guid> idSetter, Func<Object, Guid> idGetter)
        {
            Guid id = idGetter.Invoke(item);
            if (id == null)
            {
                id = Guid.NewGuid();
                idSetter.Invoke(item, id);
            }
            ConcurrentDictionary<Guid, Object> data = storage[clazz];
            data[id] = item;
            return item;
        }


        public void Delete(Guid id)
        {
            ConcurrentDictionary<Guid, Object> data = storage[clazz];
            object value;
            data.TryRemove(id, out value);
        }
    }
}
